File: AST/AST.cpp
class AstNodeFactory final {
  public:
    AstNodeFactory(AstValueFactory* ast_value_factory, Zone* zone)
      : zone_(zone),
        ast_value_factory_(ast_value_factory),
        empty_statement_(new (zone) class EmptyStatement()),
        this_expression_(new (zone) class ThisExpression()),
        failure_expression_(new (zone) class FailureExpression()) {}
    Zone* zone() const { return zone_; }
    VariableProxy* NewVariableProxy(const AstRawString* name,
                                    VariableKind variable_kind,
                                    int start_position = kNoSourcePosition) {
      return new (zone_) VariableProxy(name, variable_kind, start_position);
    }
    VariableDeclaration* NewVariableDeclaration(int pos) {
      return new (zone_) VariableDeclaration(pos);
    }
  private:
    // This zone may be deallocated upon returning from parsing a function body
    // which we can guarantee is not going to be compiled or have its AST
    // inspected.
    // See ParseFunctionLiteral in parser.cc for preconditions.
    Zone* zone_;
    AstValueFactory* ast_value_factory_;
    class EmptyStatement* empty_statement_;
    class ThisExpression* this_expression_;
    class FailureExpression* failure_expression_;
};

// ZoneObject is an abstraction that helps define classes of objects
// allocated in the Zone. Use it as a base class; see ast.h.
class ZoneObject {};

// The AST refers to variables via VariableProxies - placeholders for the actual
// variables. Variables themselves are never directly referred to from the AST,
// they are maintained by scopes, and referred to from VariableProxies and Slots
// after binding and variable allocation.
class Variable final : public ZoneObject {};

class AstNode: public ZoneObject {
  protected:
    uint32_t bit_field_;
    static const uint8_t kNextBitFieldIndex = NodeTypeField::kNext;

    AstNode(int position, NodeType type) : position_(position), bit_field_(NodeTypeField::encode(type)) {}
  private:
    // Hidden to prevent accidental usage. It would have to load the
    // current zone from the TLS.
    void* operator new(size_t size);

    int position_;
    class NodeTypeField : public BitField<NodeType, 0, 6> {};
}

class Expression : public AstNode {
  public:
    enum Context {
      // Not assigned a context yet, or else will not be visited during
      // code generation.
      kUninitialized,
      // Evaluated for its side effects.
      kEffect,
      // Evaluated for its value (and side effects).
      kValue,
      // Evaluated for control flow (and side effects).
      kTest
    };
  protected:
    Expression(int pos, NodeType type) : AstNode(pos, type) {}
};

class Declaration : public AstNode {
  public:
    typedef base::ThreadedList<Declaration> List;

    Variable* var() const { return var_; }
    void set_var(Variable* var) { var_ = var; }

  protected:
    Declaration(int pos, NodeType type) : AstNode(pos, type), next_(nullptr) {}

  private:
    Variable* var_;
    // Declarations list threaded through the declarations.
    Declaration** next() { return &next_; }
    Declaration* next_;
    friend List;
    friend base::ThreadedListTraits<Declaration>;
};

class VariableDeclaration : public Declaration {
 public:
  inline NestedVariableDeclaration* AsNested();

 private:
  friend class AstNodeFactory;

  class IsNestedField
      : public BitField<bool, Declaration::kNextBitFieldIndex, 1> {};

 protected:
  explicit VariableDeclaration(int pos, bool is_nested = false)
      : Declaration(pos, kVariableDeclaration) {
    bit_field_ = IsNestedField::update(bit_field_, is_nested);
  }

  static const uint8_t kNextBitFieldIndex = IsNestedField::kNext;
};

// The AST refers to variables via VariableProxies - placeholders for the actual
// variables. Variables themselves are never directly referred to from the AST,
// they are maintained by scopes, and referred to from VariableProxies and Slots
// after binding and variable allocation.
class Variable final : public ZoneObject {
  public:
    Variable(Scope* scope, const AstRawString* name, VariableMode mode,
            VariableKind kind, InitializationFlag initialization_flag,
            MaybeAssignedFlag maybe_assigned_flag = kNotAssigned)
        : scope_(scope),
          name_(name),
          local_if_not_shadowed_(nullptr),
          next_(nullptr),
          index_(-1),
          initializer_position_(kNoSourcePosition),
          bit_field_(MaybeAssignedFlagField::encode(maybe_assigned_flag) |
                    InitializationFlagField::encode(initialization_flag) |
                    VariableModeField::encode(mode) |
                    IsUsedField::encode(false) |
                    ForceContextAllocationField::encode(false) |
                    ForceHoleInitializationField::encode(false) |
                    LocationField::encode(VariableLocation::UNALLOCATED) |
                    VariableKindField::encode(kind)) {
      // Var declared variables never need initialization.
      DCHECK(!(mode == VariableMode::kVar &&
              initialization_flag == kNeedsInitialization));
    }
    bool is_assigned() const { return IsAssignedField::decode(bit_field_); }
    void set_is_assigned() {
      bit_field_ = IsAssignedField::update(bit_field_, true);
      if (is_resolved()) {
        var()->set_maybe_assigned();
      }
    }

    bool is_resolved() const { return IsResolvedField::decode(bit_field_); }
    void set_is_resolved() {
      bit_field_ = IsResolvedField::update(bit_field_, true);
    }
    bool is_used() { return IsUsedField::decode(bit_field_); }
    void set_is_used() { bit_field_ = IsUsedField::update(bit_field_, true); }
    MaybeAssignedFlag maybe_assigned() const {
      return MaybeAssignedFlagField::decode(bit_field_);
    }
    void set_maybe_assigned() {
      bit_field_ = MaybeAssignedFlagField::update(bit_field_, kMaybeAssigned);
    }
    void set_initializer_position(int pos) { initializer_position_ = pos; }
  private:
    Scope* scope_;
    const AstRawString* name_;
    // If this field is set, this variable references the stored locally bound
    // variable, but it might be shadowed by variable bindings introduced by
    // sloppy 'eval' calls between the reference scope (inclusive) and the
    // binding scope (exclusive).
    Variable* local_if_not_shadowed_;
    Variable* next_;
    int index_;
    int initializer_position_;
    uint16_t bit_field_;
}

enum InitializationFlag : uint8_t { kNeedsInitialization, kCreatedInitialized };

class VariableProxy final : public Expression {
  public:
    static InitializationFlag DefaultInitializationFlag(VariableMode mode) {
      return mode == VariableMode::kVar ? kCreatedInitialized : kNeedsInitialization;
    }
    Variable* var() const {
      return var_;
    }
    void set_var(Variable* v) {
      var_ = v;
    }
    void set_is_resolved() {
      bit_field_ = IsResolvedField::update(bit_field_, true);
    }
    bool is_assigned() const { return IsAssignedField::decode(bit_field_); }
    BindTo(Variable* var) {
      DCHECK_EQ(raw_name(), var->raw_name());
      set_var(var);
      set_is_resolved();
      var->set_is_used();
      if (is_assigned()) var->set_maybe_assigned();
    }
  private:
    VariableProxy(const AstRawString* name, VariableKind variable_kind, int start_position)
      : Expression(start_position, kVariableProxy),
        raw_name_(name),
        next_unresolved_(nullptr) {
      bit_field_ |= IsAssignedField::encode(false) |
                    IsResolvedField::encode(false) |
                    IsRemovedFromUnresolvedField::encode(false) |
                    HoleCheckModeField::encode(HoleCheckMode::kElided);
    }
    VariableProxy** next() { return &next_unresolved_; }
    VariableProxy* next_unresolved_;
    union {
      const AstRawString* raw_name_;  // if !is_resolved_
      Variable* var_;                 // if is_resolved_
    };
};

File: JS/ParseStatementList/ast/Ast.js
import { 
  kVar,
  kVariableDeclaration,
  kNoSourcePosition,

  kSmi,
  kHeapNumber,
  kBigInt,
  kString,
  kSymbol,
  kBoolean,
  kUndefined,
  kNull,
  kTheHole,

  kLiteral,
  kAssignment,
  kCompoundAssignment,
  kExpressionStatement,

  kVariableProxy,

  TARGET_FOR_ANONYMOUS,
  TARGET_FOR_NAMED_ONLY,

  kExpression,
  kBlock,
} from "../base/Const";

import {
  NodeTypeField,
  BreakableTypeField,
  IgnoreCompletionField,
  IsLabeledField,
  TokenField,
} from '../base/Util';

export class AstNodeFactory {
  NewVariableProxy(name, variable_kind, start_position = kNoSourcePosition) {
    return new VariableProxy(name, variable_kind, start_position);
  }
  NewVariableDeclaration(pos) {
    return new VariableDeclaration(pos);
  }

  /**
   * 生成字面量对象
   * 源码这里用的是函数重载 由于有严格的类型形参 所以很舒适
   * JS无法做到完美模拟 这里手动将对应类型的枚举传入构造函数
   * @returns {Literal} 字面量类
   */
  NewNullLiteral(pos) {
    return new Literal(kNull, null, pos);
  }
  NewBooleanLiteral(b, pos) {
    return new Literal(kBoolean, b, pos);
  }
  NewSmiLiteral(number, pos) {
    return new Literal(kSmi, number, pos);
  }
  // TODO
  // NewNumberLiteral() {}
  NewBigIntLiteral(bigint, pos) {
    return new Literal(kBigInt, bigint, pos);
  }
  NewStringLiteral(string, pos) {
    return new Literal(kString, string, pos);
  }

  /**
   * 声明+赋值语句
   * @param {Token} op 赋值类型
   * @param {Expression} target 变量名
   * @param {Expression} value 变量值
   * @param {Number} pos 位置
   * @returns {Assignment}
   */
  NewAssignment(op, target, value, pos) {
    if (op !== 'Token::INIT' && target.IsVariableProxy()) target.AsVariableProxy().set_is_assigned();
    if (op === 'Token::ASSIGN' || op === 'Token::INIT') return new Assignment(kAssignment, op, target, value, pos);
    else return new CompoundAssignment(op, target, value, pos, this.NewBinaryOperation(BinaryOpForAssignment(op), target, value, pos + 1));
  }
  NewExpressionStatement(expression, pos) {
    return new ExpressionStatement(expression, pos);
  }
  /**
   * 该方法有重载
   */
  NewBlock(ignore_completion_value, statements) {
    // 构造参数在这里基本上毫无意义
    let result = new Block();
    result.InitializeStatements(statements, null);
    return result;
  }
}

class AstNode {
  constructor(position, type) {
    this.position_ = position;
    this.bit_field_ = NodeTypeField.encode(type);
  }
  IsVariableProxy() { return this.node_type() === kVariableProxy; }
  node_type() { return NodeTypeField.decode(this.bit_field_); }
}

class Statement extends AstNode {
  constructor(position, type) {
    super(position, type);
    this.kNextBitFieldIndex = 6;
  }
}

class ExpressionStatement extends Statement {
  constructor(expression, pos) {
    super(pos, kExpressionStatement);
    this.expression_ = expression;
  }
  set_expression(e) {
    this.expression_ = e;
  }
}

/**
 * break语句除了for、while等关键词的break操作
 * 还包括了foo: if (b) { f(); break foo; }此类语法
 */
class BreakableStatement extends Statement {
  constructor(breakable_type, position, type) {
    super(position, type);
    this.bit_field_ |= BreakableTypeField.encode(breakable_type);
  }
}

class Block extends BreakableStatement {
  constructor(zone = null, labels = null, capacity = 0, ignore_completion_value = true) {
    super(TARGET_FOR_NAMED_ONLY, kNoSourcePosition, kBlock);
    this.statement_ = null;
    this.scope_ = null;
    this.bit_field_ |= IgnoreCompletionField.encode(1) | IsLabeledField.encode(0);
  }
  // 这里用的内存拷贝
  InitializeStatements(statements, zone = null) {
    this.statement_ = statements;
  }
}

class LabeledBlock extends Block {
  // 这里构造函数重载了 capacity参数后移了一位 去掉了zone参数
  constructor(labels, ignore_completion_value, capacity = 0) {
    super(labels, capacity, ignore_completion_value);
    this.labels_ = labels;
  }
}

class Declaration extends AstNode {
  constructor(pos, type) {
    super(pos, type);
    this.next_ = null;
    this.var_ = null;
  }
  var() { return this.var_; }
  set_var(v) { this.var_ = v; }

  // Declarations list threaded through the declarations.
  // Declaration** next() { return &next_; }
  // Declaration* next_;
}

class VariableDeclaration extends Declaration {
  constructor(pos, is_nested = false) {
    super(pos, kVariableDeclaration);
    this.bit_field_ = NodeTypeField.update(this.bit_field_, is_nested);
  }
}

export class Expression extends AstNode {
  constructor(pos, type) {
    super(pos, type);
  }
}

class Assignment extends Expression {
  constructor(node_type, op, target, value, pos) {
    super(pos, node_type);
    this.target_ = target;
    this.value_ = value;
    this.bit_field_ |= TokenField.encode(op);
  }
}

class CompoundAssignment extends Assignment {
  constructor(op, target, value, pos, binary_operation) {
    super(kCompoundAssignment, op, target, value, pos);
    // BinaryOperation* binary_operation_;
    this.binary_operation_ = binary_operation;
  }
}

export class VariableProxy extends Expression {
  constructor(name, variable_kind, start_position) {
    super(start_position, kVariableProxy);
    this.raw_name_ = name;
    this.next_unresolved_ = null;

    this.var_ = null;

    this.bit_field_ |= 0;
  }
  set_var(v) { this.var_ = v; }
  set_is_resolved() { this.bit_field_ = NodeTypeField.update(this.bit_field_, true); }
  is_assigned() { return NodeTypeField.decode(this.bit_field_); }

  BindTo(variable) {
    this.set_var(variable);
    this.set_is_resolved();
    variable.set_is_used();
    if (this.is_assigned()) variable.set_maybe_assigned();
  }
}

/**
 * 字面量类
 */
class Literal extends Expression{
  /**
   * 源码构造函数只用了2个参数 由于重载的存在 所以类型已经确定了
   * JS手动传进来
   * @param {Enumerator} type 枚举类型
   * @param {any} val 字面量的值
   * @param {Number} pos 位置
   */
  constructor(type, val, pos) {
    super(pos, kLiteral);
    /**
     * 每一种字面量只会有唯一的值 不可能同时是字符串与数字
     * 因此源码在这里用了union来优化内存的使用
     * 包括了6个字面量变量string_、smi_、number_、symbol_、bigint_、boolean_
     */
    this.type = type;
    /**
     * JS既没有函数重载(参数不同可以模拟 类型实在无力) 也没有union数据结构
     * 连枚举也没有(实际上用ES6的Symbol模拟枚举是一个可行性方案 但是成本略高)
     * 因此用一个数组来保存值 用val()来获取对应的值
     */
    this.val = new Array(6).fill(null);
    this.val[type] = val;
    
    this.bit_field_ = NodeTypeField.update(this.bit_field_, type);
  }
  // add
  val() {
    return this.val[this.type];
  }
}

/**
 * 这个比较特殊
 * 放最下面了
 */
// The AST refers to variables via VariableProxies - placeholders for the actual
// variables. Variables themselves are never directly referred to from the AST,
// they are maintained by scopes, and referred to from VariableProxies and Slots
// after binding and variable allocation.
const kNotAssigned = 0;
const kMaybeAssigned = 1;

const kNeedsInitialization = 0;
const kCreatedInitialized = 1;
class ZoneObject {}
export class Variable extends ZoneObject {
  constructor(scope, name, mode, kind, initialization_flag, maybe_assigned_flag = kNotAssigned) {
    super();
    this.scope_ = scope;
    this.name_ = name;
    this.local_if_not_shadowed_ = false;
    this.next_ = null;
    this.index_ = -1;
    this.initializer_position_ = kNoSourcePosition;
    this.bit_field_ = 0; // TODO
  }
  set_is_used() { this.bit_field_ = NodeTypeField.update(this.bit_field_, true); }
  set_maybe_assigned() { this.bit_field_ = NodeTypeField.update(this.bit_field_, kMaybeAssigned); }

  static DefaultInitializationFlag(mode) { return mode === kVar ? kCreatedInitialized : kNeedsInitialization; }
};